README.md
# Python SDK for Klaviyo's Beta API

- SDK version: 1.1.1
- API revision: 2022-12-15.pre

## Helpful Resources

- [Beta API Reference](https://developers.klaviyo.com/en/v2022-12-15.pre/reference)
- [Latest API Guides](https://developers.klaviyo.com/en/docs)
- [Postman Workspace](https://www.postman.com/klaviyo/workspace/klaviyo-developers)
- [Interactive Guide (Jupyter Notebook)](https://github.com/klaviyo-labs/klaviyo-api-guides)

## Design & Approach

NOTE: This SDK only reflects the endpoints currently in Beta. Once endpoints are promoted from Beta to General Availability (GA), they will be removed from the next version of this SDK, and the version of this SDK corresponding to that Beta release will be deprecated.

This SDK is a thin wrapper around our current Beta API. See our API Reference for full documentation on behavior.

This SDK exactly mirrors the organization and naming convention of the above language-agnostic resources, with a few namespace changes to make it Pythonic (details in Appendix).

## Organization

This SDK is organized into the following resources:



- Campaigns



- Data_Privacy



- Flows



- Lists



- Segments



- Tags



## Installation

You can install this library using [our pip package here](https://pypi.org/project/klaviyo-api-beta/).

Depending on your system configuration, you will need to run *one* of the following shell commands:

```bash
pip install klaviyo-api-beta
```

OR 

```bash
pip3 install klaviyo-api-beta
```

## Usage Example

### To instantiate the client

NOTE: 
* The SDK retries on resolvable errors, namely: rate limits (common) and server errors (rare).
* The keyword arguments define some advanced settings; the example is populated with the default values.
* `max_delay` denotes total delay (in seconds) across all attempts.

```python
from klaviyo_api_beta import KlaviyoAPIBeta

klaviyo_beta = KlaviyoAPIBeta("YOUR API KEY HERE", max_delay=60, max_retries=3, test_host=None)
```

### Example request

```python
klaviyo_beta.Tags.get_tags() 
```

## Error Handling

This SDK throws an `ApiException` error when the server returns a non-`2XX` response. 

An `ApiException` consists of the following attributes:

* `status` : `int`
* `reason` : `str`
* `body` : `bytes`
    * this can be decoded into a native python dictionary as follows:
        ```python
        # to decode to a dictionary
        import json
        BODY_DICT = json.loads(YOUR_EXCEPTION.body)

        # to decode to a string
        BODY_STRING = YOUR_EXCEPTION.body.decode('utf-8')
        ```
* `headers` : [class 'urllib3._collections.HTTPHeaderDict'](https://urllib3.readthedocs.io/en/stable/user-guide.html?highlight=httpheaderdict#response-content)
    * This can be interacted with as a normal dictionary:
        * ex:
            ```
            date = YOUR_EXCEPTION.headers['Date']
            keys = YOUR_EXCEPTION.headers.keys()
            values = YOUR_EXCEPTION.headers.values()
            ```

## Important Notes

- The main difference between this SDK and the language-agnostic API Docs that the below endpoints link to is that this SDK automatically adds the `revision` header corresponding to the SDK version.
- Organization: Resource groups and operation_ids are listed below in alphabetical order, first by Resource name, then by **OpenAPI Summary**. Operation summaries are those listed in the right side bar of the [API Reference](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_events).
- For example values / data types, as well as whether parameters are required/optional, please reference the corresponding API Reference link.
- Some keyword args may potentially be required for the API call to succeed, the linked API docs are the source of truth regarding which keyword params are required.
- JSON payloads should be passed in as native python dictionaries.
- You can override the client private key by passing in an optional `api_key` keyword arg to any API call that takes a private key. As a reminder: do NOT do this client-side/onsite.

# Comprehensive list of Operations & Parameters





## Campaigns

#### [Assign Campaign Message Template](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/assign_campaign_message_template)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Campaigns.assign_campaign_message_template(body)
```




#### [Create Campaign](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/create_campaign)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Campaigns.create_campaign(body)
```




#### [Create Campaign Clone](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/create_campaign_clone)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Campaigns.create_campaign_clone(body)
```




#### [Create Campaign Send Job](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/create_campaign_send_job)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Campaigns.create_campaign_send_job(body)
```




#### [Delete Campaign](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/delete_campaign)

```python
## Positional Arguments

# id | str

klaviyo_beta.Campaigns.delete_campaign(id)
```




#### [Get Campaign](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_campaign)

```python
## Positional Arguments

# id | str

## Keyword Arguments

# fields_campaign | [str]
# fields_tag | [str]
# include | [str]

klaviyo_beta.Campaigns.get_campaign(id, fields_campaign=fields_campaign, fields_tag=fields_tag, include=include)
```




#### [Get Campaign Message](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_campaign_message)

```python
## Positional Arguments

# id | str

## Keyword Arguments

# fields_campaign_message | [str]

klaviyo_beta.Campaigns.get_campaign_message(id, fields_campaign_message=fields_campaign_message)
```




#### [Get Campaign Relationships](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_campaign_relationships)

```python
## Positional Arguments

# id | str
# related_resource | str

klaviyo_beta.Campaigns.get_campaign_relationships(id, related_resource)
```




#### [Get Campaign Send Job](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_campaign_send_job)

```python
## Positional Arguments

# id | str

## Keyword Arguments

# fields_campaign_send_job | [str]

klaviyo_beta.Campaigns.get_campaign_send_job(id, fields_campaign_send_job=fields_campaign_send_job)
```




#### [Get Campaign Tags](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_campaign_tags)

```python
## Positional Arguments

# campaign_id | str

## Keyword Arguments

# fields_tag | [str]

klaviyo_beta.Campaigns.get_campaign_tags(campaign_id, fields_tag=fields_tag)
```




#### [Get Campaigns](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_campaigns)

```python

## Keyword Arguments

# fields_campaign | [str]
# fields_tag | [str]
# filter | str
# include | [str]
# sort | str

klaviyo_beta.Campaigns.get_campaigns(fields_campaign=fields_campaign, fields_tag=fields_tag, filter=filter, include=include, sort=sort)
```




#### [Update Campaign](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/update_campaign)

```python
## Positional Arguments

# id | str
# body | dict

klaviyo_beta.Campaigns.update_campaign(id, body)
```




#### [Update Campaign Message](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/update_campaign_message)

```python
## Positional Arguments

# id | str
# body | dict

klaviyo_beta.Campaigns.update_campaign_message(id, body)
```




#### [Update Campaign Send Job](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/update_campaign_send_job)

```python
## Positional Arguments

# id | str
# body | dict

klaviyo_beta.Campaigns.update_campaign_send_job(id, body)
```






## Data_Privacy

#### [Request Profile Deletion](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/request_profile_deletion)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Data_Privacy.request_profile_deletion(body)
```






## Flows

#### [Get Flow Tags](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_flow_tags)

```python
## Positional Arguments

# flow_id | str

## Keyword Arguments

# fields_tag | [str]

klaviyo_beta.Flows.get_flow_tags(flow_id, fields_tag=fields_tag)
```






## Lists

#### [Get List Tags](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_list_tags)

```python
## Positional Arguments

# list_id | str

## Keyword Arguments

# fields_tag | [str]

klaviyo_beta.Lists.get_list_tags(list_id, fields_tag=fields_tag)
```






## Segments

#### [Get Segment Tags](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_segment_tags)

```python
## Positional Arguments

# segment_id | str

## Keyword Arguments

# fields_tag | [str]

klaviyo_beta.Segments.get_segment_tags(segment_id, fields_tag=fields_tag)
```






## Tags

#### [Create Tag](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/create_tag)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Tags.create_tag(body)
```




#### [Create Tag Group](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/create_tag_group)

```python
## Positional Arguments

# body | dict

klaviyo_beta.Tags.create_tag_group(body)
```




#### [Create Tag Relationships](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/create_tag_relationships)

```python
## Positional Arguments

# id | str
# related_resource | str
# body | dict

klaviyo_beta.Tags.create_tag_relationships(id, related_resource, body)
```




#### [Delete Tag](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/delete_tag)

```python
## Positional Arguments

# id | str

klaviyo_beta.Tags.delete_tag(id)
```




#### [Delete Tag Group](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/delete_tag_group)

```python
## Positional Arguments

# id | str

klaviyo_beta.Tags.delete_tag_group(id)
```




#### [Delete Tag Relationships](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/delete_tag_relationships)

```python
## Positional Arguments

# id | str
# related_resource | str
# body | dict

klaviyo_beta.Tags.delete_tag_relationships(id, related_resource, body)
```




#### [Get Tag](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_tag)

```python
## Positional Arguments

# id | str

## Keyword Arguments

# fields_tag | [str]

klaviyo_beta.Tags.get_tag(id, fields_tag=fields_tag)
```




#### [Get Tag Group](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_tag_group)

```python
## Positional Arguments

# id | str

## Keyword Arguments

# fields_tag_group | [str]

klaviyo_beta.Tags.get_tag_group(id, fields_tag_group=fields_tag_group)
```




#### [Get Tag Group Relationships](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_tag_group_relationships)

```python
## Positional Arguments

# id | str
# related_resource | str

klaviyo_beta.Tags.get_tag_group_relationships(id, related_resource)
```




#### [Get Tag Groups](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_tag_groups)

```python

## Keyword Arguments

# fields_tag_group | [str]
# filter | str
# page_cursor | str
# sort | str

klaviyo_beta.Tags.get_tag_groups(fields_tag_group=fields_tag_group, filter=filter, page_cursor=page_cursor, sort=sort)
```




#### [Get Tag Relationships](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_tag_relationships)

```python
## Positional Arguments

# id | str
# related_resource | str

klaviyo_beta.Tags.get_tag_relationships(id, related_resource)
```




#### [Get Tags](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/get_tags)

```python

## Keyword Arguments

# fields_tag | [str]
# filter | str
# page_cursor | str
# sort | str

klaviyo_beta.Tags.get_tags(fields_tag=fields_tag, filter=filter, page_cursor=page_cursor, sort=sort)
```




#### [Update Tag](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/update_tag)

```python
## Positional Arguments

# id | str
# body | dict

klaviyo_beta.Tags.update_tag(id, body)
```




#### [Update Tag Group](https://developers.klaviyo.com/en/v2022-12-15.pre/reference/update_tag_group)

```python
## Positional Arguments

# id | str
# body | dict

klaviyo_beta.Tags.update_tag_group(id, body)
```




# Appendix

## Global Keyword Args

NOTE: These are arguments that you can apply to any endpoint call, and which are unique to the SDK

We currently support the following global keyword args:
- `api_key` : use this to override the client-level api_key which you define upon client instantiation

## Refresher on catching exceptions:

```python
try:
    YOUR_CALL
except Exception as e:
    print(e.status)
    print(e.reason)
    print(e.body)
    print(e.headers)
```

## Parameters & Arguments

The parameters follow the same naming conventions as the resource groups and operations.

We stick to the following convention for parameters/arguments

1. All parameters are passed as function args.
2. All query and path params that are tagged as `required` in the docs are passed as positional args.
3. All optional query params are passed as keyword args.
4. Where applicable, the `body` param is passed in as a positional arg, and is expected to be a native python dictionary. Within that dictionary, refer to the API docs to see which fields are required/optional, along with valid values.
4. There is no need to pass in your private `api_key` for any operations, as it is defined upon client instantiation; public key is still required where applicable. However, you can pass in an optional `api_key` kwarg to override the client private key for a specific call (REMINDER: don't do this client-side).

## Namespace

In the interest of making the SDK Pythonic, we made the following namespace changes *relative* to the language agnostic resources up top (API Docs, Guides, etc).

- Resource names use Title + Snake Casing, (e.g. `Data_Privacy`)
- function names and parameter names remain unchanged and use snake case (e.g. `get_metrics`, and `profile_id`)

CHANGELOG.md
# Changelog

NOTE: For more granular API-specific changes, please see our [API Changelog](https://developers.klaviyo.com/en/docs/changelog_)

## 1.0.0

  * Initial release

  NOTE: This SDK only reflects the endpoints currently in Beta. Once endpoints are promoted from Beta to General Availability (GA), they will be removed from the next version of this SDK, and the version of this SDK corresponding to that Beta release will be deprecated.

  - Namespace changes
    - Pypi package name: `klaviyo-api` → `klaviyo-api-beta`
    - Module name: `klaviyo_api` → `klaviyo_api_beta`
    - client name: `KlaviyoAPI` → `KlaviyoAPIBeta`
    - Client variable name in readme examples: `klaviyo` → `klaviyo_beta`
  - New resources and endpoints: see [API Changelog](https://developers.klaviyo.com/en/docs/changelog_) for full details

## 1.1.0

  - Added Campaigns endpoints


openapi_client/api/flows_api.py
"""
    Klaviyo API (Beta)

    The Klaviyo REST API. Please visit https://developers.klaviyo.com for more details.  # noqa: E501

    The version of the OpenAPI document: 2022-12-15.pre
    Contact: developers@klaviyo.com
    Generated by: https://openapi-generator.tech
"""


import re  # noqa: F401
import sys  # noqa: F401

from openapi_client.api_client import ApiClient, Endpoint as _Endpoint
from openapi_client.model_utils import (  # noqa: F401
    check_allowed_values,
    check_validations,
    date,
    datetime,
    file_type,
    none_type,
    validate_and_convert_types
)
from openapi_client.model.get_campaigns4_xx_response import GetCampaigns4XXResponse


class FlowsApi(object):
    """NOTE: This class is auto generated by OpenAPI Generator
    Ref: https://openapi-generator.tech

    Do not edit the class manually.
    """

    def __init__(self, api_client=None):
        if api_client is None:
            api_client = ApiClient()
        self.api_client = api_client
        self.get_flow_tags_endpoint = _Endpoint(
            settings={
                'response_type': ({str: (bool, date, datetime, dict, float, int, list, str, none_type)},),
                'auth': [
                    'Klaviyo-API-Key'
                ],
                'endpoint_path': '/api/flows/{flow_id}/tags/',
                'operation_id': 'get_flow_tags',
                'http_method': 'GET',
                'servers': None,
            },
            params_map={
                'all': [
                    'flow_id',
                    'fields_tag',
                ],
                'required': [
                    'flow_id',
                ],
                'nullable': [
                ],
                'enum': [
                    'fields_tag',
                ],
                'validation': [
                ]
            },
            root_map={
                'validations': {
                },
                'allowed_values': {
                    ('fields_tag',): {

                        "NAME": "name"
                    },
                },
                'openapi_types': {
                    'flow_id':
                        (str,),
                    'fields_tag':
                        ([str],),
                },
                'attribute_map': {
                    'flow_id': 'flow_id',
                    'fields_tag': 'fields[tag]',
                },
                'location_map': {
                    'flow_id': 'path',
                    'fields_tag': 'query',
                },
                'collection_format_map': {
                    'fields_tag': 'csv',
                }
            },
            headers_map={
                'accept': [
                    'application/json'
                ],
                'content_type': [],
            },
            api_client=api_client
        )

    def get_flow_tags(
        self,
        flow_id,
        **kwargs
    ):
        """Get Flow Tags  # noqa: E501

        Return all tags associated with the given flow ID.<br><br>*Rate limits*:<br>Burst: `3/s`<br>Steady: `60/m`  **Scopes:** `Flows Read` `Tags Read`  # noqa: E501
        This method makes a synchronous HTTP request by default. To make an
        asynchronous HTTP request, please pass async_req=True

        >>> thread = api.get_flow_tags(flow_id, async_req=True)
        >>> result = thread.get()

        Args:
            flow_id (str): 

        Keyword Args:
            fields_tag ([str]): For more information please visit https://developers.klaviyo.com/en/v/reference/api-overview#sparse-fieldsets. [optional]
            _return_http_data_only (bool): response data without head status
                code and headers. Default is True.
            _preload_content (bool): if False, the urllib3.HTTPResponse object
                will be returned without reading/decoding response data.
                Default is True.
            _request_timeout (int/float/tuple): timeout setting for this request. If
                one number provided, it will be total request timeout. It can also
                be a pair (tuple) of (connection, read) timeouts.
                Default is None.
            _check_input_type (bool): specifies if type checking
                should be done one the data sent to the server.
                Default is True.
            _check_return_type (bool): specifies if type checking
                should be done one the data received from the server.
                Default is True.
            _spec_property_naming (bool): True if the variable names in the input data
                are serialized names, as specified in the OpenAPI document.
                False if the variable names in the input data
                are pythonic names, e.g. snake case (default)
            _content_type (str/None): force body content-type.
                Default is None and content-type will be predicted by allowed
                content-types and body.
            _host_index (int/None): specifies the index of the server
                that we want to use.
                Default is read from the configuration.
            _request_auths (list): set to override the auth_settings for an a single
                request; this effectively ignores the authentication
                in the spec for a single request.
                Default is None
            async_req (bool): execute request asynchronously

        Returns:
            {str: (bool, date, datetime, dict, float, int, list, str, none_type)}
                If the method is called asynchronously, returns the request
                thread.
        """
        kwargs['async_req'] = kwargs.get(
            'async_req', False
        )
        kwargs['_return_http_data_only'] = kwargs.get(
            '_return_http_data_only', True
        )
        kwargs['_preload_content'] = kwargs.get(
            '_preload_content', True
        )
        kwargs['_request_timeout'] = kwargs.get(
            '_request_timeout', None
        )
        kwargs['_check_input_type'] = kwargs.get(
            '_check_input_type', True
        )
        kwargs['_check_return_type'] = kwargs.get(
            '_check_return_type', True
        )
        kwargs['_spec_property_naming'] = kwargs.get(
            '_spec_property_naming', False
        )
        kwargs['_content_type'] = kwargs.get(
            '_content_type')
        kwargs['_host_index'] = kwargs.get('_host_index')
        kwargs['_request_auths'] = kwargs.get('_request_auths', None)
        kwargs['flow_id'] = \
            flow_id
        return self.get_flow_tags_endpoint.call_with_http_info(**kwargs)

